clarity_ext/dilution_strategies.py

```diff
@@ -0,0 +1,137 @@
+                transfer.sample_volume = None
+                transfer.buffer_volume = None
+                transfer.has_to_evaporate = None
+
+
+class PoolConcentrationCalc:
+    """
+    Comments on key variable calculations:
+    * sample_volume,
+      depends on the number of samples contained within the pool, otherwise
+      like a one-to-one dilution
+    * buffer_volume,
+      based on requested volume and all sample volumes within the pool.
+      The buffer volume is specified at one single row in the driver
+      file, chosen at random among samples within the pool.
+    * has_to_evaporate,
+      same as previous, based on requested volume and all sample
+      volumes within the pool
+    * scaled_up,
+      the scale up factor is based on the min sample volume within the pool.
+      If volumes are to be scaled up, all sample volumes and the buffer volume
+      have to be scaled up.
+    """
+
+    def __init__(self):
+        pass
+
+    def calculate_transfer_volumes(self, transfers=None, scale_up_low_volumes=None):
+        transfers = sorted(transfers, key=lambda t: t.target_aliquot_name)
+        grouped_transfers = groupby(
+            transfers, key=lambda t: t.target_aliquot_name)
+        for key, group in grouped_transfers:
+            self._calc_single_pool(list(group), scale_up_low_volumes)
+
+    def _calc_single_pool(self, transfers, scale_up_low_volumes):
+        """
+        Calculate transfer volumes for a single pool
+        :param transfers: A list of transfers associated with a single pool
+        :return:
+        """
+        for transfer in transfers:
+            try:
+                transfer.sample_volume = \
+                    transfer.requested_concentration * transfer.requested_volume / \
+                    transfer.source_concentration / len(transfers)
+                transfer.buffer_volume = 0
+            except (TypeError, ZeroDivisionError):
+                transfer.sample_volume = 0
+                transfer.buffer_volume = 0
+                transfer.has_to_evaporate = False
+
+        t = transfers[0]
+        t.buffer_volume = max(t.requested_volume -
+                              sum([tt.sample_volume for tt in transfers]), 0)
+
+        min_sample_volume = min(map(lambda t: t.sample_volume, transfers))
+        for t in transfers:
+            t.has_to_evaporate = t.requested_volume - \
+                sum([tt.sample_volume for tt in transfers]) < 0
+            try:
+                if scale_up_low_volumes is True and min_sample_volume < ROBOT_MIN_VOLUME:
+                    scale_factor = float(ROBOT_MIN_VOLUME/min_sample_volume)
+                    t.sample_volume *= scale_factor
+                    t.buffer_volume *= scale_factor
+                    t.scaled_up = True
+            except ZeroDivisionError:
+                # Zero sample volume indicates an error that should be caught
+                # later by validation
+                pass
```

Comment:
I'm always afraid of catch-all exception handlers. (I understand that in this case it will work, since we currently know how the class should be used - i.e. this call should be followed by validation. But if someone refactors this later, that might not be the case anymore.)